### PR TITLE
Implement `/tryfix` feature that auto fixes links that has redirections

### PR DESCRIPTION
**Pull request by Youssef1313**
Assume you have a file called `A.md` which is renamed to `B.md` and a redirection is added. If there are links that refers to `A.md`, they should be updated to `B.md`. All that will be needed to fix this is commenting on the PR with `/tryfix` and wait for an automated commit. This will require a new yml file:

```yml
name: MSDocs build verifier - tryfix
on:
  issue_comment:
    types: [created]

jobs:
  tryfix:
    if: github.event.issue.pull_request != '' && contains(github.event.comment.body, '/tryfix')
    name: Try fix build
    runs-on: ubuntu-latest
    env:
      IS_TRY_FIX: true # differentiates /tryfix from the validation-only run.
      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    steps:
    - name: Checkout the repository
      uses: actions/checkout@a81bbbf8298c0fa03ea29cdc473d45769f953675 #@v2
    - name: Checkout Pull Request
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      run: |
        PR_URL="${{ github.event.issue.pull_request.url }}"
        PR_NUM=${PR_URL##*/}
        echo "Checking out from PR #$PR_NUM based on URL: $PR_URL"
        hub pr checkout $PR_NUM
    - name: Tryfix
      uses: Youssef1313/docs-actions/actions/docs-verifier@docs-actions

    - name: Push changes
      run: |
          git config --global user.name github-actions
          git config --global user.email github-actions@github.com
          git commit -am "Automated: Fix links"
          git push
```

@gewarren @BillWagner @IEvangelist Can we merge this into a feature branch instead of `main`? Then we can test it in a dummy PR in dotnet/docs without breaking anything (in case this is breaking something)

### CODE REVIEW

File: actions/docs-verifier/Directory.Build.props [hunk 1]
```diff
@@ -3,15 +3,15 @@
...
-    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
-    <AnalysisLevel>latest</AnalysisLevel>
+    <AnalysisLevel>latest-All</AnalysisLevel>
     <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="5.0.3" />
+    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="7.0.0-preview1.22068.2" />
```
Comment (IEvangelist) [inline actions/docs-verifier/Directory.Build.props:14]: I think we should probably stay on .NET 6 (LTS), as it is stable.

File: .github/workflows/build-docs-verifier.yml [hunk 2]
```diff
@@ -15,6 +15,8 @@ jobs:
           fetch-depth: 0
       - name: Setup .NET
         uses: actions/setup-dotnet@v1.7.2
+        with:
+          dotnet-version: '6.0.100'
```
Comment (IEvangelist) [inline .github/workflows/build-docs-verifier.yml:19]:
```suggestion
          dotnet-version: '6.0.x'
```

File: actions/docs-verifier/Dockerfile [hunk 3]
```diff
@@ -1,11 +1,11 @@
-FROM mcr.microsoft.com/dotnet/sdk:5.0.200 AS build
+FROM mcr.microsoft.com/dotnet/sdk:6.0.100 AS build
```
Comment (IEvangelist) [inline actions/docs-verifier/Dockerfile:1]:
```suggestion
FROM mcr.microsoft.com/dotnet/sdk:6.0 AS build
```

File: actions/docs-verifier/global.json [hunk 4]
```diff
@@ -1,5 +1,5 @@
 {
   "sdk": {
-    "version": "5.0.200"
+    "version": "6.0.100"
```
Comment (IEvangelist) [inline actions/docs-verifier/global.json:3]:
```suggestion
    "version": "6.0.200"
```

File: actions/docs-verifier/src/ActionRunner/Program.cs [hunk 5]
```diff
@@ -45,6 +39,60 @@
     }
 }
 
+if (Environment.GetEnvironmentVariable("IS_TRY_FIX") == "true")
```
Comment (IEvangelist) [inline actions/docs-verifier/src/ActionRunner/Program.cs:42]:
```suggestion
if (Environment.GetEnvironmentVariable("IS_TRY_FIX") is "true")
```

File: actions/docs-verifier/src/ActionRunner/Program.cs [hunk 6]
```diff
@@ -45,6 +39,60 @@
...
+        if (redirection is { RedirectUrl: string redirectUrl } && redirectUrl.StartsWith(dotnetPrefix, StringComparison.Ordinal))
+        {
+            string newAbsolutePath = "docs/" + redirectUrl[dotnetPrefix.Length..];
+            if (File.Exists(newAbsolutePath + ".md"))
+            {
+                newAbsolutePath += ".md";
+            }
+            else if (File.Exists(newAbsolutePath + ".yml"))
+            {
+                newAbsolutePath += ".yml";
```
Comment (IEvangelist) [inline actions/docs-verifier/src/ActionRunner/Program.cs:60]:
```suggestion
            string newAbsolutePath = $"docs/{redirectUrl[dotnetPrefix.Length..]}";
            if (File.Exists($"{newAbsolutePath}.md"))
            {
                newAbsolutePath += ".md";
            }
            else if (File.Exists($"{newAbsolutePath}.yml"))
            {
                newAbsolutePath += ".yml";
```

**Review by IEvangelist — COMMENTED** [review]
This looks good to me, but I don't think that having it in a branch would work - I thought we disabled GitHub Actions on feature branches, but I could be wrong.

File: actions/docs-verifier/Directory.Build.props (hunk 1 above)
Comment (Youssef1313) [inline actions/docs-verifier/Directory.Build.props:14]:
```suggestion
```

Comment (Youssef1313) [reply actions/docs-verifier/Directory.Build.props:14]: @IEvangelist It has a false positive for top-level statements. I could suppress the warning, but I preferred a preview version as long as it doesn't have any problems.

**Comment by Youssef1313** [comment]
> This looks good to me, but I don't think that having it in a branch would work - I thought we disabled GitHub Actions on feature branches, but I could be wrong.

Recently, a change was tested in https://github.com/dotnet/docs/pull/27387 via a branch in this repository. Has something changed after that?

File: actions/docs-verifier/Directory.Build.props (hunk 1 above)
Comment (IEvangelist) [reply actions/docs-verifier/Directory.Build.props:14]: I'm fine with it, but let's add a comment inline stating that so it doesn't come off as a mistake.

```xml
<!-- This preview version fixes a false positive bug with top-level statements. -->
```

**Review by IEvangelist — APPROVED** [review]

**Comment by IEvangelist** [comment]
> > This looks good to me, but I don't think that having it in a branch would work - I thought we disabled GitHub Actions on feature branches, but I could be wrong.
> 
> Recently, a change was tested in [dotnet/docs#27387](https://github.com/dotnet/docs/pull/27387) via a branch in this repository. Has something changed after that?

Ah, yes. You're right let's do the test then.

File: actions/docs-verifier/Directory.Build.props (hunk 1 above)
Comment (Youssef1313) [reply actions/docs-verifier/Directory.Build.props:14]:
```suggestion
    <!-- This preview version fixes a CA1812 (Avoid uninstantiated internal classes) false positive bug with top-level statements. -->
    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="7.0.0-preview1.22068.2" />
```

**Comment by Youssef1313** [comment]
@IEvangelist Can you create the branch? Thanks!

**Comment by IEvangelist** [comment]
> @IEvangelist Can you create the branch? Thanks!

Hey @Youssef1313, sure - see here: https://github.com/dotnet/docs/tree/youssef-testing

**Comment by Youssef1313** [comment]
@IEvangelist The needed branch is in this repo not dotnet/docs

**Comment by IEvangelist** [comment]
> @IEvangelist The needed branch is in this repo not dotnet/docs

Oops, I'm sorry. I thought you meant for over in docs - here you go. https://github.com/dotnet/docs-tools/tree/youssef-testing

**Comment by Youssef1313** [comment]
No problem! Rebased the PR on the new branch. Thanks.

**Comment by Youssef1313** [comment]
@IEvangelist Anything left to merge this?

**Pull request merged**